src/data/finsupp/lex.lean

```diff
@@ -0,0 +1,90 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.pi.lex
+import data.finsupp.order
+
+/-!
+# Lexicographic order on finitely supported functions
+
+This file defines the lexicographic order on `finsupp`.
+-/
+
+variables {α N : Type*}
+
+namespace finsupp
+
+lemma filter_ne_eq_empty_iff [decidable_eq α] [decidable_eq N] [has_zero N] {f g : α →₀ N} :
+  (f.support ∪ g.support).filter (λ a, f a ≠ g a) = ∅ ↔ f = g :=
+begin
+  refine ⟨λ h, _, λ h, h ▸ by simp⟩,
+  ext a,
+  contrapose h,
+  refine finset.ne_empty_of_mem (_ : a ∈ _),
+  simp only [ne.def, finset.mem_filter, finset.mem_union, finsupp.mem_support_iff],
+  exact ⟨not_and_distrib.1 $ mt (and_imp.2 eq.substr) (ne.symm h), h⟩,
+end
+
+section N_has_zero
+variables [has_zero N]
+
+/-- The lexicographic relation on `α →₀ N`, where `α` is ordered by `r`,
+  and `N` is ordered by `s`. -/
+protected def lex (r : α → α → Prop) (s : N → N → Prop) (x y : α →₀ N) : Prop :=
+pi.lex r (λ _, s) x y
+
+instance [has_lt α] [has_lt N] : has_lt (lex (α →₀ N)) := ⟨finsupp.lex (<) (<)⟩
+
+instance lex.is_strict_order [linear_order α] [partial_order N] :
+  is_strict_order (lex (α →₀ N)) (<) :=
+let i : is_strict_order (lex (α → N)) (<) := pi.lex.is_strict_order in
+{ irrefl := to_lex.surjective.forall.2 $ λ a, @irrefl _ _ i.to_is_irrefl a,
+  trans := to_lex.surjective.forall₃.2 $ λ a b c, @trans _ _ i.to_is_trans a b c }
+
+variables [linear_order α]
+
+/--  The partial order on `finsupp`s obtained by the lexicographic ordering.
+`finsupp.lex.linear_order` is the proof that this partial order is in fact linear. -/
+instance lex.partial_order [partial_order N] : partial_order (lex (α →₀ N)) :=
+partial_order.lift (λ x, to_lex ⇑(of_lex x)) finsupp.coe_fn_injective--fun_like.coe_injective
+
+variable [linear_order N]
+/--  The linear order on `finsupp`s obtained by the lexicographic ordering. -/
+noncomputable instance lex.linear_order : linear_order (lex (α →₀ N)) :=
+{ le_total := to_lex.surjective.forall₂.2 $ λ f g : lex (α →₀ N), begin
```

Comment:
```suggestion
{ le_total := to_lex.surjective.forall₂.2 $ λ f g, begin
```

Reply:
This annotation is abusing defeq, it should be `f g : α →₀ N` (but the best way to not be wrong is to not make the claim in the first place)

Reply:
Ok, I got confused by the comment about the `f = j` not making sense and seeing that `f` had what I thought was the "wrong" type.  Anyway, with the `forall\2` and without the type-ascription, everything compiles on my computer!